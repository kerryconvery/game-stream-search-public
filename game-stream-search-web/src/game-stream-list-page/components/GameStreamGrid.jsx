import React from 'react';
import { shape, string, number, bool, arrayOf, func } from 'prop-types';
import { makeStyles, styled } from '@material-ui/core/styles';
import Skeleton from '@material-ui/lab/Skeleton';
import IconButton from '@material-ui/core/IconButton';
import PlayCircleOutlineIcon from '@material-ui/icons/PlayCircleOutline';
import Avatar from '@material-ui/core/Avatar';
import Tooltip from '@material-ui/core/Tooltip';
import Link from '../../shared-components/Link';

const useStreamTileStyles = makeStyles(() => ({
  root: {
    'font-family': 'Helvetica',
    color: '#231E1D',
    position: 'relative'
  },
  streamTitle: {
    display: '-webkit-box',
    '-webkit-box-orient': 'vertical',
    '-webkit-line-clamp': '2',
    fontSize: '16px',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
    whiteSpace: 'normal',
    marginTop: 0,
    marginBottom: '0.25rem',
  },
  detailsContainer: {
    display: 'flex',
    flexDirection: 'row',
    paddingTop: '0.5rem',
  },
  streamDetails: {
    paddingLeft: '0.5rem',
  },
  streamSubDetails: {
    color: '#606060',
    fontSize: '14px',
    ' & > div': {
      paddingBottom: '0.2rem'
    }
  },
  playButton: {
    color: 'white',
    opacity: 0,
    position: 'absolute',
    left: 0, 
    right: 0,
    top: '10%',
    marginLeft: 'auto',
    marginRight: 'auto', 
    width: '100px', /* Need a specific value to work */
    transition: 'opacity 0.5s ease',
  },
  playButtonIcon: {
    fontSize: 100,
    color: 'white',
  },
}));


const StreamTile = ({
  streamTitle,
  streamThumbnailUrl,
  streamUrl,
  streamerName,
  streamerAvatarUrl,
  streamPlatformName,
  views,
  onStreamOpened, }) => {

  const classes = useStreamTileStyles();

  return (
    <div className={classes.root}>
      <Link href={streamUrl} onClick={onStreamOpened} target='_blank'>
        <IconButton size="medium" className={classes.playButton}>
          <PlayCircleOutlineIcon className={classes.playButtonIcon} />
        </IconButton >
        <img src={streamThumbnailUrl} width={320} height={180} />
        <div className={classes.detailsContainer}>
          <Avatar src={streamerAvatarUrl} />
          <div className={classes.streamDetails}>
            <Tooltip title={streamTitle}>
              <h3 className={classes.streamTitle}>{streamTitle}</h3>
            </Tooltip>
            <div className={classes.streamSubDetails}>
              <div>{streamerName}</div>
              <span>{streamPlatformName}</span>
              <span> â€¢ </span>
              <span>{`${views} viewers`}</span>
            </div>
          </div>
        </div>
      </Link>
    </div>
  )
}

StreamTile.propTypes = {
  streamTitle: string.isRequired,
  streamerName: string.isRequired,
  streamThumbnailUrl: string.isRequired,
  streamUrl: string.isRequired,
  streamerAvatarUrl: string.isRequired,
  streamPlatformName: string.isRequired,
  views: number.isRequired,
  onStreamOpened: func,
}

const useLoadingTileStyles = makeStyles(() => ({
  detailsContainer: {
    display: 'flex',
    flexDirection: 'row',
    paddingTop: '0.5rem',
    width: '100%',
  },
  details: {
    paddingLeft: '0.5rem',
    width: '80%',
    height: '80%',
  },
  subDetails: {
    width: '50%',
  },
}));

const LoadingTile = () => {
  const classes = useLoadingTileStyles();

  return (
    <div data-testid='stream-loading-tile'>
      <Skeleton variant='rect' height='60%' animation='wave' />
      <div className={classes.detailsContainer}>
        <Skeleton variant='circle' width={50} height={50} animation='wave' />
        <div className={classes.details}>
          <Skeleton variant='text' animation='wave' />
          <div className={classes.subDetails}>
            <Skeleton variant='text' animation='wave' />
            <Skeleton variant='text' animation='wave' />
          </div>
        </div>
      </div>
    </div>
  )
}

const Grid = styled('div')({
  display: 'flex',
  flexWrap: 'wrap',
  flexDirection: 'row',
  justifyContent: 'flex-start',
  overflow: 'hidden',
  width: '100%'
});

const GridTile = styled('div')({
  width:'20rem',
  height:'19rem',
  paddingLeft: '10px',
  paddingRight: '10px',
});

const GameStreamGrid = ({ streams, isLoading, numberOfLoadingTiles, onStreamOpened }) => {
  const streamTitle = streams.map((stream, index) => (
    <GridTile key={index} >
      <StreamTile {...stream} onStreamOpened={() => onStreamOpened(stream)} />
    </GridTile>
  ))

  const loadingTiles = [];

  if (isLoading) {
    for (let index = 0; index < numberOfLoadingTiles; index++) {
      loadingTiles.push(
        <GridTile key={index}>
          <LoadingTile />
        </GridTile>
      )
    }
  }

  return (
    <Grid>
      {streamTitle}
      {loadingTiles}
    </Grid>
  )
}

GameStreamGrid.propTypes = {
  streams: arrayOf(shape({
    streamTitle: string.isRequired,
    streamerName: string.isRequired,
    streamThumbnailUrl: string.isRequired,
    streamUrl: string.isRequired,
    streamerAvatarUrl: string.isRequired,
    streamPlatformName: string.isRequired,
    views: number.isRequired,
  })),
  isLoading: bool,
  numberOfLoadingTiles: number.isRequired,
  onStreamOpened: func,
}

GameStreamGrid.defaultProps = {
  streams: [],
  isLoading: false,
}

export default GameStreamGrid;
