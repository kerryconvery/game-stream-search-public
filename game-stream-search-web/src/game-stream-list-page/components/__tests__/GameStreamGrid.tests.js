import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import GameStreamGrid from '../GameStreamGrid';

describe('Game stream grid', () => {
  it('should render stream tiles without loading tiles', async () => {
    const streams = [
      {
        streamTitle: 'test stream A',
        streamerName: 'test streamer',
        streamThumbnailUrl: 'http://test.stream.thumbnail.url',
        streamUrl: 'http://test.stream.url',
        streamerAvatarUrl: 'http://test.channel.thumbnail.url',
        streamPlatformName: 'test platform',
        views: 100
      },
      {
        streamTitle: 'test stream B',
        streamerName: 'test streamer',
        streamThumbnailUrl: 'http://test.stream.thumbnail.url',
        streamUrl: 'http://test.stream.url',
        streamerAvatarUrl: 'http://test.channel.thumbnail.url',
        streamPlatformName: 'test platform',
        views: 10
      },
    ]

    const { container } = render(<GameStreamGrid streams={streams} numberOfLoadingTiles={1} />)

    const streamTiles = await waitFor(() => screen.getAllByText('test streamer'));
    const loadingTiles = screen.queryAllByTestId('stream-loading-tile');
    
    expect(streamTiles.length).toEqual(2);
    expect(loadingTiles.length).toEqual(0);
    expect(container.firstChild).toMatchSnapshot();
  })

  it('should render loading tiles when there are no streams', async () => {
    const { container } = render(<GameStreamGrid isLoading numberOfLoadingTiles={2} />)

    const loadingTiles = await waitFor(() => screen.getAllByTestId('stream-loading-tile'));
    
    expect(loadingTiles.length).toEqual(2);
    expect(container.firstChild).toMatchSnapshot();
  })

  it('should render stream and loading tiles', async () => {
    const streams = [
      {
        streamTitle: 'test stream A',
        streamerName: 'test streamer',
        streamThumbnailUrl: 'http://test.stream.thumbnail.url',
        streamUrl: 'http://test.stream.url',
        streamerAvatarUrl: 'http://test.channel.thumbnail.url',
        streamPlatformName: 'test platform',
        views: 100
      },
    ]

    const { container } = render(<GameStreamGrid streams={streams} isLoading numberOfLoadingTiles={1} />)

    const streamTiles = await waitFor(() => screen.getAllByText('test streamer'));
    const loadingTiles = screen.getAllByTestId('stream-loading-tile');
    
    expect(streamTiles.length).toEqual(1);
    expect(loadingTiles.length).toEqual(1);
  })
})


