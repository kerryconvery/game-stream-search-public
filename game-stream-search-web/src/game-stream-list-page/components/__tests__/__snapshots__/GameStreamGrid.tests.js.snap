// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game stream grid should render loading tiles when there are no streams 1`] = `
<div
  class="div-root-11"
>
  <div
    class="div-root-12"
  >
    <div
      data-testid="stream-loading-tile"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-wave"
        style="height: 60%;"
      />
      <div
        class="makeStyles-detailsContainer-13"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-wave"
          style="width: 50px; height: 50px;"
        />
        <div
          class="makeStyles-details-14"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          />
          <div
            class="makeStyles-subDetails-15"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
            />
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="div-root-12"
  >
    <div
      data-testid="stream-loading-tile"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-rect MuiSkeleton-wave"
        style="height: 60%;"
      />
      <div
        class="makeStyles-detailsContainer-13"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-circle MuiSkeleton-wave"
          style="width: 50px; height: 50px;"
        />
        <div
          class="makeStyles-details-14"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
          />
          <div
            class="makeStyles-subDetails-15"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
            />
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Game stream grid should render stream tiles without loading tiles 1`] = `
<div
  class="div-root-1"
>
  <div
    class="div-root-2"
  >
    <div
      class="makeStyles-root-3"
      data-test-id="stream-tile"
    >
      <a
        class="a-root-10"
        href="http://test.stream.url"
        target="_blank"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root makeStyles-playButton-8"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-playButtonIcon-9"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 16.5l6-4.5-6-4.5v9zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <img
          height="180"
          src="http://test.stream.thumbnail.url"
          width="320"
        />
        <div
          class="makeStyles-detailsContainer-5"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circle"
          >
            <img
              class="MuiAvatar-img"
              src="http://test.channel.thumbnail.url"
            />
          </div>
          <div
            class="makeStyles-streamDetails-6"
          >
            <h3
              class="makeStyles-streamTitle-4"
              title="test stream A"
            >
              test stream A
            </h3>
            <div
              class="makeStyles-streamSubDetails-7"
            >
              <div>
                test streamer
              </div>
              <span>
                test platform
              </span>
              <span>
                 • 
              </span>
              <span>
                100 viewers
              </span>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
  <div
    class="div-root-2"
  >
    <div
      class="makeStyles-root-3"
      data-test-id="stream-tile"
    >
      <a
        class="a-root-10"
        href="http://test.stream.url"
        target="_blank"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root makeStyles-playButton-8"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-playButtonIcon-9"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 16.5l6-4.5-6-4.5v9zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <img
          height="180"
          src="http://test.stream.thumbnail.url"
          width="320"
        />
        <div
          class="makeStyles-detailsContainer-5"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circle"
          >
            <img
              class="MuiAvatar-img"
              src="http://test.channel.thumbnail.url"
            />
          </div>
          <div
            class="makeStyles-streamDetails-6"
          >
            <h3
              class="makeStyles-streamTitle-4"
              title="test stream B"
            >
              test stream B
            </h3>
            <div
              class="makeStyles-streamSubDetails-7"
            >
              <div>
                test streamer
              </div>
              <span>
                test platform
              </span>
              <span>
                 • 
              </span>
              <span>
                10 viewers
              </span>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;
